### PR TITLE
task finished

### PR DESCRIPTION
**Pull request by Fowlart**

### CODE REVIEW

File: src/main/scala/GeoHashAppender.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import ch.hsr.geohash.GeoHash
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+import java.util.Objects
+
+class GeoHashAppender extends java.io.Serializable {
+
+  def addGeoHashToHotels(cleansedHotelDf: DataFrame, ss: SparkSession): DataFrame = {
```
Comment (xGxAxBxAx) [inline src/main/scala/GeoHashAppender.scala:8]: Better change to Datasets

Comment (xGxAxBxAx) [reply src/main/scala/GeoHashAppender.scala:8]: (rowHotels: Dataset[Hotel], ss: SparkSession): Dataset[ConvertedHotel] =

File: src/main/scala/ConvertedHotelWithGeoHash.scala [hunk 2]
```diff
@@ -0,0 +1,9 @@
+case class ConvertedHotelWithGeoHash(id: BigInt,
+                                     name: String,
+                                     country: String,
+                                     city: String,
+                                     address: String,
+                                     latitude: String,
+                                     longitude: String,
+                                     geohash: String
+                                    ) {}
```
Comment (xGxAxBxAx) [inline src/main/scala/ConvertedHotelWithGeoHash.scala:9]: Remove "{}"

File: src/main/scala/ConvertedHotel.scala [hunk 3]
```diff
@@ -0,0 +1,7 @@
+case class ConvertedHotel(Id: BigInt,
+                          Name: String,
+                          Country: String,
+                          City: String,
+                          Address: String,
+                          Latitude: String,
+                          Longitude: String) {}
```
Comment (xGxAxBxAx) [inline src/main/scala/ConvertedHotel.scala:7]: 1. Names to lowercase.
2. Remove "{}"

File: src/main/scala/Hotel.scala [hunk 4]
```diff
@@ -0,0 +1,7 @@
+case class Hotel(Id: BigInt,
+                 Name: String,
+                 Country: String,
+                 City: String,
+                 Address: String,
+                 Latitude: Option[String],
+                 Longitude: Option[String]) {}
```
Comment (xGxAxBxAx) [inline src/main/scala/Hotel.scala:7]: 1. Names to lowercase.
2. Remove "{}"

File: src/main/scala/ScMain.scala [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  weatherWithGeoHashDF.write.mode(SaveMode.Overwrite).parquet("results/weather_with_geoHash")
+
+  /** =Task 3: =
+   * Left join weather and hotels data by generated 4-characters
+   * geohash (avoid data multiplication and make you job idempotent) */
+  val joinCondition = cleansedHotelWithGeoHashDF.col("GeoHash") === weatherWithGeoHashDF.col("weather_GeoHash")
+  val resultDf = cleansedHotelWithGeoHashDF.join(weatherWithGeoHashDF, joinCondition, "left_outer")
+  resultDf.repartition(1)
+    .write.mode(SaveMode.Overwrite)
+    .mode(SaveMode.Overwrite)
```
Comment (xGxAxBxAx) [inline src/main/scala/ScMain.scala:59]: option "mode" duplicate

File: src/main/scala/WeatherRecord.scala [hunk 6]
```diff
@@ -0,0 +1,9 @@
+case class WeatherRecord(lng: String,
+                         lat: String,
+                         avg_tmpr_f: String,
+                         avg_tmpr_c: String,
+                         wthr_date: String,
+                         year: String,
+                         month: String,
+                         day: String
+                        ) {}
```
Comment (xGxAxBxAx) [inline src/main/scala/WeatherRecord.scala:9]: Remove "{}"

File: src/main/scala/WeatherRecordWithGeoHash.scala [hunk 7]
```diff
@@ -0,0 +1,10 @@
+case class WeatherRecordWithGeoHash(lng: String,
+                                    lat: String,
+                                    avg_tmpr_f: String,
+                                    avg_tmpr_c: String,
+                                    wthr_date: String,
+                                    year: String,
+                                    month: String,
+                                    day: String,
+                                    geohash: String
+                                   ) {}
```
Comment (xGxAxBxAx) [inline src/main/scala/WeatherRecordWithGeoHash.scala:10]: Remove "{}"

File: src/main/scala/GeoHashAppender.scala [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    import ss.implicits._
+
+    cleansedHotelDf.as[ConvertedHotel].rdd.map(location => {
```
Comment (xGxAxBxAx) [inline src/main/scala/GeoHashAppender.scala:11]: not necessary ".rdd"

File: src/main/scala/GeoHashAppender.scala [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        location.Id,
+        location.Name,
+        location.Country,
+        location.City,
+        location.Address,
+        location.Latitude,
+        location.Longitude,
+        GeoHash.geoHashStringWithCharacterPrecision(latitude, longitude, 4)
+      )
+    }).toDF("Id", "Name", "Country", "City", "Address", "Latitude", "Longitude", "GeoHash")
```
Comment (xGxAxBxAx) [inline src/main/scala/GeoHashAppender.scala:26]: Using Dataset you can remove that

File: src/main/scala/GeoHashAppender.scala [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+
+  def addGeoHashToWeather(weatherDf: DataFrame, ss: SparkSession): DataFrame = {
+    import ss.implicits._
+
+    weatherDf.as[WeatherRecord].rdd
```
Comment (xGxAxBxAx) [inline src/main/scala/GeoHashAppender.scala:33]: not necessary ".rdd"

File: src/main/scala/GeoHashAppender.scala [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        location.Address,
+        location.Latitude,
+        location.Longitude,
+        GeoHash.geoHashStringWithCharacterPrecision(latitude, longitude, 4)
+      )
+    }).toDF("Id", "Name", "Country", "City", "Address", "Latitude", "Longitude", "GeoHash")
+  }
+
+
+  def addGeoHashToWeather(weatherDf: DataFrame, ss: SparkSession): DataFrame = {
```
Comment (xGxAxBxAx) [inline src/main/scala/GeoHashAppender.scala:30]: better rewrite to Dataset

File: src/main/scala/HotelFilter.scala [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import com.byteowls.jopencage.JOpenCageGeocoder
+import com.byteowls.jopencage.model.JOpenCageForwardRequest
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+class HotelFilter extends java.io.Serializable {
+
+  def retrieveCoordinates(street: String,
+                          city: String,
+                          countryCode: String) = {
+    val jOpenCageGeocoder = new JOpenCageGeocoder("5f189b638d024369ac7318e4eedb9a5c")
```
Comment (xGxAxBxAx) [inline src/main/scala/HotelFilter.scala:11]: maybe create constant for "5f189b638d024369ac7318e4eedb9a5c"

File: src/main/scala/ScMain.scala [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    .option("header", "true")
+    .option("inferSchema", "true")
+    .csv("abfs://m06sparkbasics@bd201stacc.dfs.core.windows.net/hotels")
+
+  /** =Task 1:  =
+   * Check hotels data on incorrect (null) values (Latitude & Longitude).
+   * For incorrect values map (Latitude & Longitude) from OpenCage Geocoding
+   * API in job on fly (Via REST API). */
+  val hotelFilter = new HotelFilter
+  val cleansedHotelDf = hotelFilter.cleanAndEnhanceHotelsDf(hotelsDf, sparkSession)
```
Comment (xGxAxBxAx) [inline src/main/scala/ScMain.scala:39]: With this use, "HotelFilter" can be the object not class

File: src/main/scala/ScMain.scala [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  cleansedHotelDf.write.mode(SaveMode.Overwrite).parquet("results/cleaned_hotels")
+
+  /** =Task 2: =
+   * Generate geohash by Latitude & Longitude using one of geohash libraries (like geohash-java)
+   * with 4-characters length in extra column */
+  val geoHashAppender = new GeoHashAppender
+  val cleansedHotelWithGeoHashDF = geoHashAppender.addGeoHashToHotels(cleansedHotelDf, sparkSession)
```
Comment (xGxAxBxAx) [inline src/main/scala/ScMain.scala:46]: With this use, "GeoHashAppender" can be the object not class

File: src/main/scala/ScMain.scala [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  // weather
+  val weatherDf = sparkSession.
+    read.parquet("abfs://m06sparkbasics@bd201stacc.dfs.core.windows.net/weather")
+
+  // hotels
+  val hotelsDf = sparkSession
+    .read
+    .option("header", "true")
+    .option("inferSchema", "true")
+    .csv("abfs://m06sparkbasics@bd201stacc.dfs.core.windows.net/hotels")
```
Comment (xGxAxBxAx) [inline src/main/scala/ScMain.scala:32]: Dataset more safe and convenient than Dataframe, please rewrite to them

**Review by xGxAxBxAx — CHANGES_REQUESTED** [review]
Great for the first time. But better improve some places.

File: src/main/scala/GeoHashAppender.scala (hunk 1 above)
Comment (Fowlart) [reply src/main/scala/GeoHashAppender.scala:8]: fixed

File: src/main/scala/ConvertedHotelWithGeoHash.scala (hunk 2 above)
Comment (Fowlart) [reply src/main/scala/ConvertedHotelWithGeoHash.scala:9]: Fixed

File: src/main/scala/ConvertedHotel.scala (hunk 3 above)
Comment (Fowlart) [reply src/main/scala/ConvertedHotel.scala:7]: fixed

File: src/main/scala/Hotel.scala (hunk 4 above)
Comment (Fowlart) [reply src/main/scala/Hotel.scala:7]: Fixed

File: src/main/scala/ScMain.scala (hunk 5 above)
Comment (Fowlart) [reply src/main/scala/ScMain.scala:59]: fixed

File: src/main/scala/GeoHashAppender.scala (hunk 11 above)
Comment (Fowlart) [reply src/main/scala/GeoHashAppender.scala:30]: done

File: src/main/scala/HotelFilter.scala (hunk 12 above)
Comment (Fowlart) [reply src/main/scala/HotelFilter.scala:11]: fixed

File: src/main/scala/ScMain.scala (hunk 13 above)
Comment (Fowlart) [reply src/main/scala/ScMain.scala:39]: Fixed.

File: src/main/scala/ScMain.scala (hunk 14 above)
Comment (Fowlart) [reply src/main/scala/ScMain.scala:46]: Fixed

File: src/main/scala/ScMain.scala (hunk 15 above)
Comment (Fowlart) [reply src/main/scala/ScMain.scala:32]: Done

File: src/main/scala/WeatherRecord.scala (hunk 6 above)
Comment (Fowlart) [reply src/main/scala/WeatherRecord.scala:9]: done

File: src/main/scala/WeatherRecordWithGeoHash.scala (hunk 7 above)
Comment (Fowlart) [reply src/main/scala/WeatherRecordWithGeoHash.scala:10]: Done

**Pull request merged**